### PR TITLE
[SSD] Added SSD trainer in experimental

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Fixes # (issue)

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Adds a new experimental training loop that performs on-policy generation and fine-tunes on sampled completions, plus new CLI/eval scripts; this touches training/inference pathways and could surface edge cases around generation buffering, masking, and vLLM integration. Also makes `add_model_tags` unconditional in self-distillation trainers, which could break with non-standard model wrappers that lack this method.
> 
> **Overview**
> **Adds an experimental `SSDTrainer` implementation** (`trl.experimental.ssd`) that generates completions on-the-fly (Transformers or optional vLLM), optionally filters empty/stub outputs, reuses buffered generations across steps, and trains with standard cross-entropy on the generated completions while logging SSD-specific metrics.
> 
> Includes a new `SSDConfig`, a CLI entry script (`ssd.py`) and a LiveCodeBench v6 evaluation script (`ssd_eval.py`), plus unit tests covering prompt formats, buffering behavior, PEFT, metric logging, and key sampling hyperparameters.
> 
> Documentation is updated to surface SSD in the experimental toctree and paper index, and self-distillation trainers (`SDFTTrainer` and `BaseSelfDistillationTrainer`) now call `self.model.add_model_tags(...)` unconditionally.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1e454f5a2538eba9f167f89999f5e878b7f9abd5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: docs/source/paper_index.md [hunk 1]
```diff
@@ -1708,6 +1708,38 @@ Expected dataset columns:
...
+Simple Self-Distillation (SSD) improves code generation by sampling completions from the model at a training-time temperature and truncation configuration, then fine-tuning on those raw, unverified samples with standard cross-entropy loss. No reward model, verifier, teacher model, or reinforcement learning is needed. SSD reshapes token distributions in a context-dependent way: suppressing distractor tails at "lock" positions (where syntax leaves little ambiguity) while preserving diversity at "fork" positions (where multiple valid continuations exist).
+
+```python
+from trl.experimental.ssd import SSDConfig, SSDTrainer
+
+training_args = SSDConfig(
+    temperature=0.6,                       # Training-time sampling temperature (T_train)
+    top_k=20,                              # Training-time top-k truncation
+    top_p=0.95,                            # Training-time top-p truncation
+    num_generations=1,                     # N=1 already suffices
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: we could hardcode `num_generations` I think (remove this arg)

File: tests/experimental/test_ssd_trainer.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+        assert config.vllm_mode == "colocate"
+        assert config.vllm_model_impl == "vllm"
+
+    def test_training_with_string_prompts(self):
+        dataset = Dataset.from_dict(
+            {
+                "prompt": ["Write a function to add two numbers.", "Write a function to check if a number is prime."],
+            }
+        )
```
Comment (qgallouedec) [inline tests/experimental/test_ssd_trainer.py]: prefer using `trl-internal-testing/zen` when possible

File: trl/experimental/ssd/ssd_config.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    _VALID_DICT_FIELDS = TrainingArguments._VALID_DICT_FIELDS + ["model_init_kwargs"]
+
+    model_init_kwargs: dict[str, Any] | None = field(
+        default=None,
+        metadata={"help": "Keyword arguments for model initialization when `model` is passed as a string."},
+    )
+    remove_unused_columns: bool = field(
+        default=False,
+        metadata={"help": "Whether to drop dataset columns unused by the trainer."},
+    )
```
Comment (qgallouedec) [inline trl/experimental/ssd/ssd_config.py]: do we need this?

File: trl/experimental/ssd/ssd_config.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    ds3_gather_for_generation: bool = field(
+        default=True,
+        metadata={"help": "Whether to gather ZeRO-3 weights for generation."},
+    )
+
+    def __post_init__(self):
+        super().__post_init__()
+
+        if self.num_generations < 1:
```
Comment (qgallouedec) [inline trl/experimental/ssd/ssd_config.py]: if we remove num_generations and always use 1, it would simplify this

File: trl/experimental/ssd/ssd_trainer.py [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+            raise TypeError("The `processing_class` must be either a `PreTrainedTokenizerBase` or a `ProcessorMixin`")
+
+        if tokenizer.pad_token is None:
+            tokenizer.pad_token = tokenizer.eos_token
+
+        self.pad_token = tokenizer.pad_token
+        self.pad_token_id = tokenizer.pad_token_id
+        self.eos_token_id = tokenizer.eos_token_id
+        self.max_prompt_length = args.max_prompt_length
+        self.max_completion_length = args.max_completion_length
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/ssd/ssd_trainer.py:165]: good point, you should use something like

https://github.com/huggingface/trl/blob/ea283c6e4f47266d5b73adb705ff2e4f2c05e73c/trl/trainer/sft_trainer.py#L660-L664

File: trl/experimental/ssd/ssd_trainer.py [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+        )
+        prompt_inputs = _BaseTrainer._prepare_inputs(self, prompt_inputs)
+        prompt_ids = prompt_inputs["input_ids"]
+        prompt_mask = prompt_inputs["attention_mask"]
+
+        # Log completion statistics
+        mode = "train" if self.model.training else "eval"
+        completion_lengths = completion_mask.sum(dim=1).float()
+        agg_lengths = self.accelerator.gather(completion_lengths)
+        self._metrics[mode]["completions/mean_length"].append(agg_lengths.mean().item())
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/ssd/ssd_trainer.py:450]: yep you should include the log function like in other trainers

**Review by qgallouedec — APPROVED** [review]
cool! There are some things to improve but it looks good to me overall!

**Comment by kashif** [comment]
@qgallouedec all issues addressed

**Review by sergiopaniego — COMMENTED** [review]
great work!
I'd suggest moving the eval script to examples/scripts/evals and document that in the trainer + examples overview ([here](https://huggingface.co/docs/trl/example_overview))

wdyt @kashif?

**Comment by kashif** [comment]
ok sure! let me move it

**Comment by sergiopaniego** [comment]
forget my previous suggestion, I'd say it makes more sense to keep the examples isolated in this case since it's an experimental trainer. Instead, these examples could be documented in docs/source/ssd_trainer.md.

**Comment by kashif** [comment]
ok makes sense!

File: docs/source/ssd_trainer.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+No `privileged_context`, reward functions, or teacher model are needed.
+
+## Key hyperparameters
+
+The paper identifies the following key hyperparameters:
+
+- **`temperature`**: training-time sampling temperature (T_train). Higher values create more diverse samples but may include more noise. The paper uses T_train=0.6 with truncation.
+- **`top_k`** and **`top_p`**: training-time truncation parameters (rho_train). These suppress low-probability distractor tails during data synthesis.
+- **T_eval**: the evaluation-time decoding temperature is set independently at inference time. The paper shows that T_train and T_eval compose through an effective temperature T_eff = T_train * T_eval, with a broad optimal band.
+
```
Comment (sergiopaniego) [inline docs/source/ssd_trainer.md:64]:
```suggestion

## Example script                                                                                                            
 
Use [`trl/experimental/ssd/ssd.py`](https://github.com/huggingface/trl/blob/main/trl/experimental/ssd/ssd.py) to launch SSD training from the command line. The script supports any causal LM from the Hub, custom local datasets via `--dataset_path`,
and PEFT/LoRA via the standard `ModelConfig` flags.                                                                          
                
```bash
python trl/experimental/ssd/ssd.py \
    --model_name_or_path Qwen/Qwen3-4B-Instruct-2507 \
    --dataset_name microsoft/rStar-Coder \                                                                                   
    --dataset_config seed_sft \
    --prompt_column question \                                                                                               
    --output_dir outputs/ssd-qwen3-4b \
    --per_device_train_batch_size 1 \                                                                                        
    --gradient_accumulation_steps 32 \
    --learning_rate 5e-6 \                                                                                                   
    --lr_scheduler_type cosine \
    --max_prompt_length 1024 \
    --max_completion_length 65536 \                                                                                          
    --temperature 1.6 \
    --top_k 20 \                                                                                                             
    --top_p 0.8 \
    --num_train_epochs 1 \
    --bf16 \
    --report_to wandb

```

**Review by sergiopaniego — APPROVED** [review]
